{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "import gpplot as gpp\n",
    "import csv, requests, warnings, os, matplotlib\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "matplotlib.rc(\"pdf\", fonttype=42)\n",
    "\n",
    "gpp.set_aesthetics(palette='Set2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finds max length of list of lists\n",
    "def FindMaxLength(lst): \n",
    "    maxList = max(lst, key = lambda i: len(i)) \n",
    "    maxLength = len(maxList) \n",
    "    return maxLength\n",
    "\n",
    "# Returns reverse complement of a sequence\n",
    "def revcom(s):\n",
    "    basecomp = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A','N':'N','K':'M','M':'K','R':'Y','Y':'R','S':'S','W':'W','B':'V','V':'B','H':'D','D':'H','-':'-'}\n",
    "    letters = list(s[::-1])\n",
    "    letters = [basecomp[base] for base in letters]\n",
    "    return ''.join(letters)\n",
    "\n",
    "#Codon to amino acid letter code\n",
    "def get_codon_map():\n",
    "    codon_map = {'TTT':'F', 'TTC':'F', 'TTA':'L', 'TTG':'L', 'CTT':'L', 'CTC':'L', 'CTA':'L', 'CTG':'L', 'ATT':'I', 'ATC':'I',\n",
    "             'ATA':'I', 'ATG':'M', 'GTT':'V', 'GTC':'V', 'GTA':'V', 'GTG':'V', 'TCT':'S', 'TCC':'S', 'TCA':'S', 'TCG':'S',\n",
    "             'CCT':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P', 'ACT':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T', 'GCT':'A', 'GCC':'A',\n",
    "             'GCA':'A', 'GCG':'A', 'TAT':'Y', 'TAC':'Y', 'TAA':'*', 'TAG':'*', 'CAT':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "             'AAT':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K', 'GAT':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E', 'TGT':'C', 'TGC':'C',\n",
    "             'TGA':'*', 'TGG':'W', 'CGT':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R', 'AGT':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R',\n",
    "             'GGT':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G'}\n",
    "    return codon_map\n",
    "\n",
    "# Get information about all plates associated with specified sgrna.\n",
    "# Input: sgRNA ID, like sg16\n",
    "def get_plate_info(sg):\n",
    "    df = pd.read_excel('metafiles/Plate_guide_info_v1.xlsx')\n",
    "    plate_info = df[df['sgrna_name'] == sg]\n",
    "    return plate_info\n",
    "    \n",
    "# Get sgrna frame, reverse complement status and relevant condition IDs\n",
    "# Input: sgRNA ID, like sg16\n",
    "def get_sgrna_details(sgrna):\n",
    "    input_df = pd.read_csv('metafiles/BEV_allele_freq_input_v1.csv')\n",
    "    sg_df = input_df[input_df.sg == sgrna].values[0]\n",
    "    sgrna_frame = sg_df[5]\n",
    "    sgrna_rev_status = sg_df[6]\n",
    "    sgrna_rel_cond = sg_df[-1]\n",
    "    return sgrna_frame, sgrna_rev_status, sgrna_rel_cond\n",
    "\n",
    "# Gets wells that pass number of aligned reads in a plate\n",
    "# Input: Plate number, folder with batch files\n",
    "def get_pass_wells(plate, folder):\n",
    "    plate_folder = folder + 'CRISPRessoBatch_on_CRISPResso_batch_file_'+plate+'/'\n",
    "    files = os.listdir(plate_folder)\n",
    "    map_stats_file = [x for x in files if \"mapping_statistics\" in x][0]\n",
    "    map_stats_df = pd.read_csv(plate_folder+'/'+map_stats_file, sep='\\t')\n",
    "    pass_wells = map_stats_df[map_stats_df['READS ALIGNED'] >= 10000]\n",
    "    return pass_wells, files, plate_folder\n",
    "\n",
    "# Get relevant allele file from validation and assign colors \n",
    "# Input: sgRNA ID, relavant condition \n",
    "def get_sg_bev_file(s, rel_cond):\n",
    "    pal = sns.color_palette('Set2').as_hex()\n",
    "    sgrna_file = [x for x in os.listdir('BEV_sg_files/') if s in x][0]\n",
    "    mcl1_bev = pd.read_csv('BEV_sg_files/'+sgrna_file)\n",
    "    col = '_'.join(rel_cond.split(';'))\n",
    "    return mcl1_bev\n",
    "\n",
    "# Get alleles per well with %Reads >= 10\n",
    "# Input: row of passed wells, batch files, CRISPResso output folder for specific plate \n",
    "def get_alleles(p, files, plate_folder):\n",
    "    w_folder = [x for x in files if ('CRISPResso_on_'+p.Batch in x)&('.html' not in x)][0]\n",
    "    w_files = os.listdir(plate_folder+'/'+w_folder)\n",
    "    well = w_folder.split('_')[2]\n",
    "    allele_table_file = [x for x in w_files if (\"Alleles_frequency_table_around_sgRNA\" in x)&('.png' not in x)&('.pdf' not in x)][0]\n",
    "    allele_table = pd.read_csv(plate_folder+'/'+w_folder+'/'+allele_table_file, sep='\\t')\n",
    "    allele_table = allele_table.sort_values('%Reads', ascending=False)\n",
    "    allele_table_fil = allele_table.loc[allele_table['%Reads']>=10.0]\n",
    "    return allele_table_fil, well\n",
    "\n",
    "# Aligns WT translated sequence to allele translated sequence\n",
    "def align(wt_trans, al_trans):\n",
    "    new_al_trans = ''\n",
    "    for i,n in enumerate(al_trans):\n",
    "        if n == wt_trans[i]:\n",
    "            new_al_trans+='-'\n",
    "        else:\n",
    "            new_al_trans+=n\n",
    "    return new_al_trans\n",
    "\n",
    "# Get colors and categories for all alleles \n",
    "def get_colors(allele_table, sg_bev, well, sg_frame, sg_rev_status, w, sg, plate_no, condition):\n",
    "    wt_seq = allele_table.loc[:,'Reference_Sequence'].values[0]\n",
    "    if sg_rev_status == True:\n",
    "        wt_seq = revcom(wt_seq)\n",
    "    for k,r in allele_table.iterrows():\n",
    "        cat = ''\n",
    "        color = ''\n",
    "        seq = r.Aligned_Sequence\n",
    "        if sg_rev_status == True:\n",
    "            seq = revcom(r.Aligned_Sequence)\n",
    "        matched_row = sg_bev[sg_bev.Aligned_Sequence == seq]\n",
    "        if len(matched_row) == 1:\n",
    "            cat = matched_row.Category.values[0]\n",
    "            color = matched_row.Colors.values[0]\n",
    "            wt_trans = translate(wt_seq, sg_frame)\n",
    "            al_trans = translate(seq, sg_frame)\n",
    "            if len(wt_trans) == len(al_trans):\n",
    "                al_trans = align(wt_trans, al_trans)\n",
    "        else:\n",
    "            if r.n_deleted != 0 or r.n_inserted != 0:\n",
    "                if sg_rev_status == True:\n",
    "                    indel_wt_seq = revcom(r.Reference_Sequence)\n",
    "                    indel_al_seq = revcom(r.Aligned_Sequence)\n",
    "                else:\n",
    "                    indel_wt_seq = r.Reference_Sequence\n",
    "                    indel_al_seq = r.Aligned_Sequence\n",
    "                wt_trans = translate(indel_wt_seq, sg_frame)\n",
    "                al_trans = translate(indel_al_seq, sg_frame)\n",
    "                cat = 'Indel'\n",
    "                color = '#ffd92f'     \n",
    "            else:\n",
    "                wt_trans, al_trans, cat,color = classify_allele(r.Aligned_Sequence,r.Reference_Sequence, sg_frame, sg_rev_status)\n",
    "                matched_row_trans = sg_bev[sg_bev.Translated == al_trans]\n",
    "                if len(matched_row_trans) == 1:\n",
    "                    cat = matched_row_trans.Category.values[0]\n",
    "                    color = matched_row_trans.Colors.values[0]\n",
    "                    if len(wt_trans) == len(al_trans):\n",
    "                        al_trans = align(wt_trans, al_trans)\n",
    "                else:\n",
    "                    if len(wt_trans) == len(al_trans):\n",
    "                        al_trans = align(wt_trans, al_trans)                        \n",
    "        row = list(r)\n",
    "        row.extend([wt_trans, al_trans, cat,color,well,plate_no,condition])\n",
    "        w.writerow(row)\n",
    "    return\n",
    "\n",
    "#Classifies \"Other\" alleles\n",
    "def classify_allele(aligned_seq, ref_seq, sg_frame, sg_rev_status):\n",
    "    colors = {'Silent':'#8da0cb',\n",
    "              'Nonsense':'#ffd92f'}\n",
    "    wt_allele = ref_seq\n",
    "    if sg_rev_status == True:\n",
    "        wt_allele = revcom(wt_allele)\n",
    "    wt_trans_seq = translate(wt_allele, sg_frame)\n",
    "    cat = ''\n",
    "    color = ''\n",
    "    allele = aligned_seq\n",
    "    if sg_rev_status == True:\n",
    "        allele = revcom(allele)\n",
    "    trans_seq = translate(allele, sg_frame)\n",
    "    if trans_seq == wt_allele:\n",
    "        cat = 'Silent'\n",
    "        color = colors[cat]\n",
    "    return wt_trans_seq, trans_seq, cat, color\n",
    "\n",
    "\n",
    "#Get colors and categories for all control alleles\n",
    "def get_control_colors(allele_table, well, w, plate_no, cond):\n",
    "    wt_seq = allele_table.loc[:,'Reference_Sequence'].values[0]\n",
    "    for k,r in allele_table.iterrows():\n",
    "        cat = ''\n",
    "        color = ''\n",
    "        seq = r.Aligned_Sequence\n",
    "        if seq == wt_seq:\n",
    "            cat = 'WT'\n",
    "            color = '#e78ac3'\n",
    "        else:\n",
    "            cat = 'Other'\n",
    "            color = '#b3b3b3'\n",
    "            flag = 1\n",
    "        row = list(r)\n",
    "        row.extend(['','', cat, color, well, plate_no, cond])\n",
    "        w.writerow(row)\n",
    "    return\n",
    "\n",
    "# This function returns the tranlation of a given sequence and frame\n",
    "def translate(seq, frame):\n",
    "    codon_map = get_codon_map()\n",
    "    aa = ''\n",
    "    i = frame - 1\n",
    "    while i < len(seq):\n",
    "        substring = ''\n",
    "        while len(substring) < 3:\n",
    "            if i < len(seq):                \n",
    "                if seq[i] != '-':\n",
    "                    substring += seq[i]\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if len(substring) == 3:\n",
    "            if 'N' in substring:\n",
    "                aa = aa + '-'\n",
    "            else:\n",
    "                aa = aa + codon_map[substring]\n",
    "    return aa\n",
    "\n",
    "#Transposes a dataframe and reassigns index and columns\n",
    "def transpose_df(df):\n",
    "    df2=df.T\n",
    "    header=df2.iloc[0]\n",
    "    df2=df2[1:]\n",
    "    df2.columns=header\n",
    "    return df2\n",
    "\n",
    "#Get colors for specific categories\n",
    "def get_color_hash(color_df, cat_df):\n",
    "    color_hash = {}\n",
    "    cols = list(cat_df.columns)[1:]\n",
    "    for i,c in enumerate(cols):\n",
    "        cat_list = list(cat_df[c])\n",
    "        col_list = list(color_df[c])\n",
    "        for j,ca in enumerate(cat_list):\n",
    "            if ca not in color_hash.keys():\n",
    "                color_hash[ca] = col_list[j]\n",
    "            else:\n",
    "                continue \n",
    "    return color_hash\n",
    "\n",
    "#Generate legend for bar plots\n",
    "def get_legend(color_df, cat_df):\n",
    "    color_hash = get_color_hash(color_df, cat_df)\n",
    "    if '' in color_hash.keys():\n",
    "        del color_hash['']\n",
    "    handles = []\n",
    "    labels = {}\n",
    "    for k,v in color_hash.items():\n",
    "        if v in labels.keys():\n",
    "            labels[v] = labels[v]+'/'+k\n",
    "        else:\n",
    "            labels[v] = k\n",
    "    for k,v in labels.items():\n",
    "        handles.append(mpatches.Patch(color=k, label=v))\n",
    "    return(handles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generates aggregate alleles file for each sgRNA\n",
    "sg = ['control','sg14','sg15','sg16','sg17']\n",
    "#Specify output folder path containing CRISPResso outputs\n",
    "outputfolder = 'CRISPResso_plate_outputs/'\n",
    "for i,s in enumerate(sg):\n",
    "    print(s)\n",
    "    plate_info = get_plate_info(s)\n",
    "    if s != 'control':\n",
    "        sg_frame, sg_rev_status, sg_rel_cond = get_sgrna_details(s)\n",
    "        sg_bev = get_sg_bev_file(s,sg_rel_cond)\n",
    "    outputfile = s+'_agg_file.txt'\n",
    "    with open(outputfile, 'w') as o:\n",
    "        w = csv.writer(o, delimiter='\\t')\n",
    "        w.writerow(['Aligned_Sequence','Reference_Sequence','Unedited','n_deleted','n_inserted','n_mutated','#Reads','%Reads','Reference_translation','Aligned_translation','Category','Colors','Well','Plate','Condition'])\n",
    "        for j,p in plate_info.iterrows():\n",
    "            pass_wells, files, plate_folder = get_pass_wells(p['Plate #'], outputfolder)\n",
    "            for j,pw in pass_wells.iterrows():\n",
    "                allele_table_fil, well = get_alleles(pw, files, plate_folder)\n",
    "                if s == 'control':\n",
    "                    get_control_colors(allele_table_fil, well, w, p['Plate #'], p.condition)\n",
    "                else:\n",
    "                    get_colors(allele_table_fil, sg_bev, well, sg_frame, sg_rev_status, w, s, p['Plate #'],p.condition)       \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generates stacked bar plots for treated and untreated samples of each sgRNA\n",
    "sg = ['control','sg14','sg15','sg16','sg17']\n",
    "for i,s in enumerate(sg):\n",
    "    print(s)\n",
    "    agg_df = pd.read_csv(s+'_agg_file.txt', sep='\\t')\n",
    "    agg_df['Labels'] = agg_df.Plate+'_'+agg_df.Well\n",
    "    cond = list(set(agg_df.Condition))\n",
    "    for co in cond:\n",
    "        cond_df = agg_df[agg_df.Condition == co]\n",
    "        wells = list(set(cond_df.Labels))\n",
    "        per_reads = []\n",
    "        colors = []\n",
    "        cate = []\n",
    "        for j,w in enumerate(wells):\n",
    "            w_df = cond_df[cond_df.Labels == w]\n",
    "            reads_list = [w]\n",
    "            reads_list.extend(list(w_df['%Reads']))\n",
    "            per_reads.append(reads_list)\n",
    "            colors_list = [w]\n",
    "            colors_list.extend(list(w_df.Colors))\n",
    "            colors.append(colors_list)\n",
    "            cate_list = [w]\n",
    "            cate_list.extend(list(w_df.Category))\n",
    "            cate.append(cate_list)\n",
    "        max_len = FindMaxLength(per_reads)\n",
    "        cols = ['Well']\n",
    "        cols.extend(['Allele_'+str(x) for x in range(1,max_len)])\n",
    "        bar_df = pd.DataFrame(columns=cols)\n",
    "        bar_color_df = pd.DataFrame(columns=cols)\n",
    "        bar_category = pd.DataFrame(columns=cols)\n",
    "        for i,pr in enumerate(per_reads):\n",
    "            c = colors[i]\n",
    "            cc = cate[i]\n",
    "            pr.extend([0]*(max_len-len(pr)))\n",
    "            c.extend(['#FFFFFF']*(max_len-len(c)))\n",
    "            cc.extend(['']*(max_len-len(cc)))\n",
    "            bar_df.loc[i] = pr\n",
    "            bar_color_df.loc[i] = c\n",
    "            bar_category.loc[i] = cc\n",
    "        bar_df = bar_df.sort_values('Well', ascending=True)\n",
    "        bar_color_df = bar_color_df.sort_values('Well', ascending=True)        \n",
    "        bar_category = bar_category.sort_values('Well', ascending=True)\n",
    "        data = np.array(bar_df.iloc[:,1:].T)\n",
    "        X = np.arange(data.shape[1])\n",
    "        color_data = np.array(bar_color_df.iloc[:,1:].T)\n",
    "        fig, ax = plt.subplots(figsize=(3,2))\n",
    "        for axis in ['bottom','left']:\n",
    "              ax.spines[axis].set_linewidth(0.5)\n",
    "        for i in range(data.shape[0]):\n",
    "            plt.bar(X, data[i],\n",
    "                bottom = np.sum(data[:i], axis = 0),\n",
    "                color = color_data[i % len(color_data)],width=1.0,edgecolor='white')\n",
    "        ticklabels = [x.split('_')[1] for x in list(bar_color_df.Well)]\n",
    "        plt.xticks(ticks=X, labels=ticklabels, rotation=90, fontsize=5)\n",
    "        plt.yticks(fontsize=6)\n",
    "        plt.tick_params(axis='both', width=0.5, length=4)\n",
    "        handles = get_legend(bar_color_df, bar_category)\n",
    "        plt.legend(handles=handles, bbox_to_anchor=(1.05,1), fontsize=6)\n",
    "        t = plt.ylabel('% of reads', fontsize=7)\n",
    "        t = plt.title(s+'_'+co, fontsize=7)\n",
    "        sns.despine()  \n",
    "        plt.savefig('Figures/'+s+'/'+s+'_'+co+'_stacked_bar_plot_v1.pdf', dpi=1000,bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
