{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Edit over time plots\n",
    "\n",
    "Author: Ruth Hanna\n",
    "\n",
    "This notebook contains the code to create plots showing the % of reads containing a given edit at specified timepoints (e.g. Figure 3G). \n",
    "\n",
    "The inputs are: \n",
    "1. an input file containing the required metainformation\n",
    "2. the processed allele frequency plots, which are produced by the BEV_allele_frequencies notebook\n",
    "\n",
    "The outputs are:\n",
    "1. a pdf file containing a plot that shows the % of all reads containing the specified edit at each timepoint / drug condition\n",
    "2. a csv file containing the plotted data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "\n",
    "# Set default settings for plotting\n",
    "sns.set_style('ticks')\n",
    "sns.set_context('paper')\n",
    "mpl.rc('pdf', fonttype=42)\n",
    "mpl.rcParams['axes.linewidth'] = 0.5\n",
    "mpl.rcParams['xtick.major.width'] = 0.5\n",
    "mpl.rcParams['ytick.major.width'] = 0.5\n",
    "mpl.rcParams['xtick.major.size'] = 2\n",
    "mpl.rcParams['ytick.major.size'] = 2\n",
    "mpl.rcParams['font.sans-serif'] = \"Arial\"\n",
    "mpl.rcParams['font.family'] = \"sans-serif\"\n",
    "current_palette = sns.color_palette('Set2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "This function filters out any reads with indels, as indicated\n",
    "by the presence of '-' in either the Aligned_Sequence (deletions)\n",
    "or the Reference_Sequence (insertions)\n",
    "'''\n",
    "def filter_out_indels(row):\n",
    "    if ('-' in row['Aligned_Sequence']) or ('-' in row['Reference_Sequence']):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def make_plot_v2(df,output_name,aa,title,hue_order,xlabel='days post-transduction',ylims=(0,100),figsize=(1.2,1.8)):\n",
    "    color_nums = [7,0,1,2,3,4,5,6]\n",
    "    colors = [sns.color_palette('Set2')[n] for n in color_nums]\n",
    "    fig,ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    # Plot average of 2 replicates as bar\n",
    "    sns.barplot(x='day',y=aa,data=df,palette=colors,zorder=1,linewidth=0,hue='treatment',ci=None,hue_order=hue_order)\n",
    "    \n",
    "    # Plot individual replicates as dots\n",
    "    sns.stripplot(df['day'],df[aa],color='black',zorder=2,s=3,hue=df['treatment'],dodge=True,hue_order=hue_order)\n",
    "    sns.despine(top=True,right=True)\n",
    "    plt.xlabel(xlabel,fontsize=6)\n",
    "    \n",
    "    # Tidy up plot\n",
    "    ax.set_ylim(ylims)\n",
    "    plt.xticks(fontsize=6)\n",
    "    plt.yticks(fontsize=6)\n",
    "    ax.set_title(title,fontsize=6)\n",
    "    plt.ylabel('percentage of reads with edit',fontsize=6)\n",
    "    fig.savefig('../Data/Validation_CRISPResso_results/aa_over_time/'+output_name+'.pdf',transparent=True,bbox_inches='tight')\n",
    "    plt.close()\n",
    "    return\n",
    "\n",
    "def run_v2(row,df):\n",
    "    # Read in full (unfiltered) allele file\n",
    "    alleles = pd.read_csv('../Data/Validation_CRISPResso_results/allele_freq/'+row['sg']+'_'+str(row['primer'])+\n",
    "                          '_allele_frequency_table_around_sgRNA.csv')\n",
    "    alleles['contains_indel'] = alleles.apply(filter_out_indels,axis=1)\n",
    "    alleles = alleles.loc[alleles['contains_indel'] == False,:]\n",
    "    # Get identity of edits at specified position\n",
    "    # Handles either coding edits (pos_type == 'aa') or non-coding edits (pos_type == 'nuc')\n",
    "    if row['pos_type'] == 'aa':\n",
    "        alleles['pos'] = alleles['Translated'].apply(lambda x: x[int(row['pos'])])\n",
    "    elif row['pos_type'] == 'nuc':\n",
    "        alleles['pos'] = alleles['Aligned_Sequence'].apply(lambda x: x[int(row['pos'])])\n",
    "    else:\n",
    "        print 'pos_type must be \"aa\" or \"nuc\"'\n",
    "        return\n",
    "\n",
    "    # Assign BEVs to timepoints and treatments\n",
    "    bev_dict = {}\n",
    "    for j,r in df.iterrows():\n",
    "        bev_dict['%Reads_'+r['bev']] = (r['day'],r['treatment'],r['order'])\n",
    "        \n",
    "    # Sum up %Reads for all rows with the same value for the specified residue\n",
    "    cols = bev_dict.keys() + ['pos']\n",
    "    cols.sort()\n",
    "    grouped = alleles[cols].groupby(['pos']).sum()\n",
    "    \n",
    "    # Transpose and filter for edit of interest\n",
    "    grouped = grouped[grouped.index == row['edit_to']].transpose()\n",
    "\n",
    "    for i,col in enumerate(['day','treatment','order']):\n",
    "        grouped[col] = grouped.index.map(lambda x: bev_dict[x][i])\n",
    "    \n",
    "    # Make sure the x-axis values are in the correct order\n",
    "    grouped = grouped.sort_values(by='order')\n",
    "    \n",
    "    # Get hue order from sorted dataframe\n",
    "    hue_order = grouped.loc[:,'treatment'].drop_duplicates(keep='first').tolist()\n",
    "    output_name = row['sg'] + '_' + row['edit_name']\n",
    "    make_plot_v2(df=grouped,title=row['title'],output_name=output_name,aa=row['edit_to'],ylims=(0,row['ylim']),\n",
    "                 figsize=(row['figwidth'],1.5),xlabel=row['xlabel'],hue_order=hue_order)\n",
    "    grouped.to_csv('../Data/Validation_CRISPResso_results/aa_over_time/'+output_name+'.csv')\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sg5\n",
      "sg1\n",
      "sg12\n",
      "sg3\n",
      "sg10\n",
      "sg10\n",
      "sg3\n",
      "sg29\n",
      "sg30\n",
      "sg31\n",
      "sg32\n",
      "sg32\n",
      "sg33\n",
      "sg34\n",
      "sg35\n",
      "sg36\n",
      "sg24\n",
      "sg25\n",
      "sg26\n",
      "sg27\n",
      "sg28\n",
      "sg28\n",
      "sg28\n",
      "sg16\n",
      "sg15\n",
      "sg17\n",
      "sg21\n",
      "sg19\n",
      "sg18\n",
      "sg20\n"
     ]
    }
   ],
   "source": [
    "input_file = pd.read_csv('../Data/Validation_CRISPResso_results/aa_over_time/aa_over_time_input_v3.csv')\n",
    "sg_file = input_file[['sg','primer','pos_type','pos','edit_to','edit_name','ylim','title','figwidth','xlabel']].dropna().drop_duplicates()\n",
    "\n",
    "for i,row in sg_file.iterrows():\n",
    "    df = input_file.loc[(input_file['sg'] == row['sg']) &\n",
    "                        (input_file['primer'] == row['primer']) &\n",
    "                        (input_file['pos_type'] == row['pos_type']) &\n",
    "                        (input_file['pos'] == row['pos']) & \n",
    "                        (input_file['edit_to'] == row['edit_to']) &\n",
    "                        (input_file['edit_name'] == row['edit_name']) &\n",
    "                        (input_file['ylim'] == row['ylim']) &\n",
    "                        (input_file['title'] == row['title']) &\n",
    "                        (input_file['figwidth'] == row['figwidth']) &\n",
    "                        (input_file['xlabel'] == row['xlabel']),:]\n",
    "    print row['sg']\n",
    "    run_v2(row,df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
